### PR TITLE
refactor(docs-infra): Lazy-loads SVG icons

### PR DESCRIPTION
**Pull request by sonukapoor**
Prior to this commit, SVG icons were all loaded in the constructor
of the `CustomIconRegistry`. This commit avoids that, and loads SVG
icons on demand.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: aio/src/app/shared/custom-icon-registry.ts [hunk 1]
```diff
@@ -39,35 +39,50 @@ const DEFAULT_NS = '$$default';
  */
 @Injectable()
 export class CustomIconRegistry extends MatIconRegistry {
-  private preloadedSvgElements: SvgIconMap = {[DEFAULT_NS]: {}};
+  private cachedSvgElements: SvgIconMap = {[DEFAULT_NS]: {}};
+  private svgIcons: SvgIconInfo[];
```
Comment (gkalpak) [inline aio/src/app/shared/custom-icon-registry.ts:43]: Could we mark the constructor parameter with `private` instead (and avoid the explicit declaration/assignment)?

File: aio/src/app/shared/custom-icon-registry.ts [hunk 2]
```diff
@@ -39,35 +39,50 @@ const DEFAULT_NS = '$$default';
...
     super(http, sanitizer, document, errorHandler);
-    this.loadSvgElements(svgIcons);
+    this.svgIcons = svgIcons;
   }
 
   getNamedSvgIcon(iconName: string, namespace?: string) {
-    const nsIconMap = this.preloadedSvgElements[namespace || DEFAULT_NS];
-    const preloadedElement = nsIconMap && nsIconMap[iconName];
+    const nsIconMap = this.cachedSvgElements[namespace || DEFAULT_NS];
+    let preloadedElement: SVGElement | undefined = nsIconMap && nsIconMap[iconName];
```
Comment (gkalpak) [inline aio/src/app/shared/custom-icon-registry.ts:53]: Do we need the explicit type?

File: aio/src/app/shared/custom-icon-registry.ts [hunk 3]
```diff
@@ -39,35 +39,50 @@ const DEFAULT_NS = '$$default';
...
     return preloadedElement
         ? of(preloadedElement.cloneNode(true) as SVGElement)
         : super.getNamedSvgIcon(iconName, namespace);
   }
 
-  private loadSvgElements(svgIcons: SvgIconInfo[]) {
-    svgIcons.forEach(icon => {
-      const ns = icon.namespace || DEFAULT_NS;
-      const nsIconMap = this.preloadedSvgElements[ns] || (this.preloadedSvgElements[ns] = {});
+  private lazyloadSvgElement(iconName: string, namespace?: string) {
```
Comment (gkalpak) [inline aio/src/app/shared/custom-icon-registry.ts:63]: The `lazy` part is irrelevant here. This method loads an SVG element. Let's rename to `loadSvgElement()`.

**Review by gkalpak — CHANGES_REQUESTED** [review]
Thx, @sonukapoor! I left a couple of minor comments/suggestions. Otherwise lgtm :+1:

BTW, I believe the commit message type should be `refactor` (or potentially `perfromance`, although this change is unlikely to have any tangible performance impact).

File: aio/src/app/shared/custom-icon-registry.ts (hunk 2 above)
Comment (sonukapoor) [reply aio/src/app/shared/custom-icon-registry.ts:53]: Yes, we do. The `lazyLoadSvgElement` function can return `SVGElement` or `undefined`

File: aio/src/app/shared/custom-icon-registry.ts (hunk 1 above)
Comment (sonukapoor) [reply aio/src/app/shared/custom-icon-registry.ts:43]: Oops.. yes. Should have done that in first place.

File: aio/src/app/shared/custom-icon-registry.ts (hunk 2 above)
Comment (gkalpak) [reply aio/src/app/shared/custom-icon-registry.ts:51]: But, I mean doesn't TypeScript automatically infer the type here?
(BTW, I am always in favor of adding types to method signatures, so feel free to add an explicit return type :grin:)

Comment (sonukapoor) [reply aio/src/app/shared/custom-icon-registry.ts:51]: Unfortunately, it doesn't.

**Review by gkalpak — APPROVED** [review]
LGTM, thx :sparkles:

(For future commits, please try to follow the [commit message guidelines][1] more closely. For example, the [summary][2] must use imperative, present tense and not be capitalized.)

[1]: https://github.com/angular/angular/blob/57575e379fce9eb2367bfd48a8eb8be81f1d082c/CONTRIBUTING.md#commit
[2]: https://github.com/angular/angular/blob/57575e379fce9eb2367bfd48a8eb8be81f1d082c/CONTRIBUTING.md#summary

**Pull request closed without merging**